src/part-guide/concurrency.md

```diff
@@ -0,0 +1,175 @@
+
+From the perspective of the whole system, blocking IO in a concurrent system with threads and non-blocking IO in an async concurrent system are similar. In both cases, IO takes time and other work gets done while the IO is happening:
+- With threads, the thread doing IO requests IO from the OS, the thread is paused by the OS, other threads get work done, and when the IO is done, the OS wakes up the thread so it can continue execution with the result of the IO.
+- With async, the task doing IO requests IO from the runtime, the runtime requests IO from the OS but the OS returns control to the runtime. The runtime pauses the IO task and schedules other tasks to get work done. When the IO is done, the runtime wakes up the IO task so it can continue execution with the result of the IO.
+
+The advantage of using async IO, is that the overheads are much lower so a system can support orders of magnitude more tasks than threads. That makes async concurrency particularly well-suited for tasks with lots of users which spend a lot of time waiting for IO (if they don't spend a lot of time waiting and instead do lots of CPU-bound work, then there is not so much advantage to the low-overheads because the bottleneck will be CPU and memory resources).
+
+Threads and async are not mutually exclusive: many programs use both. Some programs have parts which are better implemented using threads and parts which are better implemented using async. For example, a database server may use async techniques to manage network communication with clients, but use OS threads for computation on data. Alternatively, a program may be written only using async concurrency, but the runtime will execute tasks on multiple threads. This is necessary for a program to make use of multiple CPU cores. We'll cover the intersection of threads and async tasks in a number of places later in the book.
+
+[^threads]: We'll start our explanation assuming a program only has a single thread, but expand on that later. There will probably be other processes running on the system, but they don't really affect how async concurrency works.
+[^other]: There are some programming languages (or even libraries) which have concurrency which is managed within the program (without the OS), but with a pre-emptive scheduler rather than relying on cooperation between threads. Go is a well-known example. These systems don't require `async` and `await` notation, but have other downsides including making interop with other languages or the OS much more difficult, and having a heavyweight runtime. Very early versions of Rust had such a system, but no traces of it remained by 1.0.
+
+
+## Concurrency and Parallelism
```

Comment:
Another thought here: the way I think about futures vs threads is that threads inherently couple concurrency and parallelism into a single construct. This is exemplified by things like hyper-threading.

Futures on the other hand intentionally decouple concurrency from parallelism. Concurrency after all is just a way of scheduling work, and with futures we don't have to make use of threads for this. This I feel like might be the most important property of futures, and I think is worth highlighting.

A framing I've started using is that tasks are not the async/await form of threads; it's more accurate to think of them as parallelizable futures. This framing does not match Tokio and `async-std`'s current task design; but both also have trouble propagating cancellation. See [`parallel_future`](https://docs.rs/parallel-future/latest/parallel_future/index.html) and [tasks are the wrong abstraction](https://blog.yoshuawuyts.com/tasks-are-the-wrong-abstraction/) for more.

Reply:
I was thinking quite a bit about this perspective, and my conclusion is that 'threads inherently couple concurrency and parallelism into a single construct' (which is the way I have been thinking as well) is not quite right. It's not inherent to threads, rather it is about the way they are conventionally used and configured. It's easy enough in most OS's to pin a thread to a CPU and/or specify which threads operate on the same cores (i.e., to separate out the parallelism), it's just that this is usually not surfaced in an ergonomic way in PLs. With futures/async, some degree of configuration is explicit in the code (e.g., join vs spawn) and some can be configured in the runtime.

> A framing I've started using is that tasks are not the async/await form of threads; ...

While I agree this is a really useful framing, given that it's not how any major runtime works today and that the reader is likely to have some background with threads, I think the 'tasks are sort of threads' is going to be a more useful way to introduce readers here. I think I'll rely more on the 'tasks are parallelisable futures' framing when I introduce join, select, etc. later in the guide